src/ng/compile.js

```diff
@@ -1215,6 +1227,24 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
 
     var SIMPLE_ATTR_NAME = /^\w/;
     var specialAttrHolder = document.createElement('div');
+
+    // The onChanges hooks should all be run together in a single digest
+    // When changes occur, the call to trigger their hooks will be added to this queue
+    var onChangesQueue;
+
+    // This function is called in a $$postDigest to trigger all the onChanges hooks in a single digest
+    function flushOnChangesQueue() {
+      // We must run this hook in an apply since the $$postDigest runs outside apply
+      $rootScope.$apply(function() {
+        for (var i = 0, ii = onChangesQueue.length; i < ii; ++i) {
```

Comment:
Good call...
All the items in the onChanges queue will be flushed before the new digest is triggered.

The sequence looks like:

* some change 1
* some change 2
* $digest
  * schedule flushOnChangesQueue
  * queue onChanges 1
  * queue onChanges 2
* $$postDigest
  * flushOnChangesQueue
    * $apply
      * trigger onChanges 1
        * some change 1a
      * trigger onChanges 2
        * some change 2a
      * $digest
        * schedule onChangesFlush
        * queue onChanges 1
        * queue onChanges 2
      * $$postDigest
        * flushOnChangesQueue
          * $apply
            * trigger onChanges 1
            * trigger onChanges 2
          * $digest

I think this means that we are OK (except for the potential for deep recursion). I'll make a test